### PR TITLE
Support for declaring/deploying contract by name in test cheatcodes

### LINKED ISSUE #925: Allow referring to contracts by contract name in `declare` and `deploy_contract` cheatcodes
e.g. `deploy_contract("main")`

https://discord.com/channels/793094838509764618/964116575387021413/1026779572487852042

### PR DESCRIPTION
**Pull request by Arcticae**
fix #925 
blocked by #988

### CODE REVIEW

**Comment by mkaput** [comment]
is there a issue that is being fixed by this? please link one

File: protostar/testing/environments/test_execution_environment.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 from .common_test_cheatcode_factory import CommonTestCheatcodeFactory
 from .execution_environment import ExecutionEnvironment
+from ...compiler import ProjectCompiler
```
Comment (mkaput) [inline protostar/testing/environments/test_execution_environment.py:21]: nitpick: wouldn't it be easier to import `protostar.compiler`?

File: tests/integration/protostar_fixture.py [hunk 2]
```diff
@@ -208,7 +205,9 @@ def format_with_output(
 
         return summary, output
 
-    def create_files(self, relative_path_str_to_file: Dict[str, Union[str, Path]]) -> None:
+    def create_files(
+        self, relative_path_str_to_file: Dict[str, Union[str, Path]]
+    ) -> None:
```
Comment (mkaput) [inline tests/integration/protostar_fixture.py:210]: nothing changes in this file except formatting?

File: protostar/testing/environments/setup_execution_environment.py [hunk 3]
```diff
@@ -10,16 +10,18 @@
 from protostar.testing.starkware.test_execution_state import TestExecutionState
 
 from .common_test_cheatcode_factory import CommonTestCheatcodeFactory
+from ...compiler import ProjectCompiler
```
Comment (kasperski95) [inline protostar/testing/environments/setup_execution_environment.py:13]: [Style] I think in this case, absolute import would be better. (I use absolute imports when importing from a different module, and relative when importing from the same module.)

File: tests/conftest.py [hunk 4]
```diff
@@ -149,3 +149,24 @@ def create_file_structure(root_path: Path, file_structure_schema: FileStructureS
...
+    yield
+    toml_path.unlink(missing_ok=True)
+    if lib_dir.is_dir():
+        lib_dir.rmdir()
+
+
+@pytest.fixture(scope="function", autouse=True)
+def fake_protostar_toml_fixture():
+    with base_protostar_toml():
+        yield
```
Comment (kasperski95) [inline tests/conftest.py:172]: [Design] Global `autouse` fixture for all e2e and integration tests seems invasive.
@mkaput WDYT?

File: protostar/testing/environments/test_execution_environment.py (hunk 1 above)
Comment (Arcticae) [reply protostar/testing/environments/test_execution_environment.py:21]: it's just the thing that IDE proposed, since it saw more relative imports. will fix.

File: protostar/testing/cheatcodes/declare_cheatcode.py [hunk 5]
```diff
@@ -59,8 +59,16 @@ def declare(
 
         return DeclaredContract(class_hash)
 
-    async def _declare_contract(self, contract_path: Path):
-        contract_class = self._starknet_compiler.compile_contract(contract_path)
+    async def _declare_contract(self, contract: str):
+        contract_identifier = (
+            Path(contract).resolve() if contract.endswith(".cairo") else contract
+        )
```
Comment (mkaput) [inline protostar/testing/cheatcodes/declare_cheatcode.py:65]: couldn't this logic be shared with migrator's cheatcodes? for example by putting it inside `ProjectCompiler` class?

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -89,9 +89,10 @@ def assert_cairo_test_cases(
 
 @pytest.fixture(name="devnet_gateway_url", scope="session")
 def devnet_gateway_url_fixture(devnet_port: int):
-    proc = run_devnet(["poetry", "run", "starknet-devnet"], devnet_port)
-    yield f"http://localhost:{devnet_port}"
-    proc.kill()
+    with base_protostar_toml():
+        proc = run_devnet(["poetry", "run", "starknet-devnet"], devnet_port)
+        yield f"http://localhost:{devnet_port}"
+        proc.kill()
```
Comment (kasperski95) [inline tests/integration/conftest.py:95]: Why is this needed?

File: protostar/testing/environments/common_test_cheatcode_factory.py [hunk 7]
```diff
@@ -25,16 +26,17 @@
 
 
 class CommonTestCheatcodeFactory(CheatcodeFactory):
-    def __init__(self, state: TestExecutionState):
+    def __init__(self, state: TestExecutionState, project_compiler: ProjectCompiler):
```
Comment (mkaput) [inline protostar/testing/environments/common_test_cheatcode_factory.py:29]: I think `project_compiler` could easily land in `TestExecutionState` and thus you could save a lot of passing. (Compiler by intuition **can** be a stateful object, for example if it has incremental cache).

The structure that was until now was totally fine IMHO.

File: tests/benchmarks/test_benchmark.py [hunk 8]
```diff
@@ -67,6 +69,20 @@ def protostar_cairo_path_fixture(request) -> Path:
...
+) -> ProjectCompiler:
+    # FIXME(arcticae): Build a project compiler with fake protostar.toml in the future if those tests are needed
+    return None  # type: ignore
+    # return ProjectCompiler(
+    #     project_root_path=project_root_path,
+    #     project_cairo_path_builder=project_cairo_path_builder,
+    #     contracts_section_loader=ProtostarContractsSection.Loader(
+    #         protostar_toml_reader
+    #     ),
+    # )
```
Comment (kasperski95) [inline tests/benchmarks/test_benchmark.py:83]: Maybe this code should be deleted. It can be always restored from the git history.

File: tests/benchmarks/test_benchmark.py [hunk 9]
```diff
@@ -67,6 +69,20 @@ def protostar_cairo_path_fixture(request) -> Path:
     return (Path(request.node.fspath).parent.parent.parent / "cairo").absolute()
 
 
+def get_test_project_compiler(
+    include_paths: Optional[List[str]] = None,
+) -> ProjectCompiler:
+    # FIXME(arcticae): Build a project compiler with fake protostar.toml in the future if those tests are needed
+    return None  # type: ignore
```
Comment (mkaput) [inline tests/benchmarks/test_benchmark.py:76]:
```suggestion
    return NotImplemented
```

File: tests/benchmarks/test_benchmark.py [hunk 10]
```diff
@@ -67,6 +69,20 @@ def protostar_cairo_path_fixture(request) -> Path:
...
+    # return ProjectCompiler(
```
Comment (mkaput) [inline tests/benchmarks/test_benchmark.py:77]: Put this beside `FIXME` comment, indented, so that TODO browsers (for example IntelliJ's one) can see this snippet.

File: tests/conftest.py [hunk 11]
```diff
@@ -149,3 +149,24 @@ def create_file_structure(root_path: Path, file_structure_schema: FileStructureS
             new_root_path = root_path / Path(path_str)
             new_root_path.mkdir()
             create_file_structure(new_root_path, file_structure_schema=composite)
+
+
+@contextmanager
+def base_protostar_toml():
```
Comment (mkaput) [inline tests/conftest.py:155]:
```suggestion
def empty_protostar_project():
```

File: tests/conftest.py [hunk 12]
```diff
@@ -149,3 +149,24 @@ def create_file_structure(root_path: Path, file_structure_schema: FileStructureS
...
+            """["protostar.project"]\nlibs_path="lib"\n["protostar.contracts"]"""
+        )
+    yield
+    toml_path.unlink(missing_ok=True)
+    if lib_dir.is_dir():
+        lib_dir.rmdir()
+
+
+@pytest.fixture(scope="function", autouse=True)
+def fake_protostar_toml_fixture():
```
Comment (mkaput) [inline tests/conftest.py:170]:
```suggestion
def empty_protostart_project_fixture():
```

File: tests/integration/conftest.py [hunk 13]
```diff
@@ -89,9 +89,10 @@ def assert_cairo_test_cases(
 
 @pytest.fixture(name="devnet_gateway_url", scope="session")
 def devnet_gateway_url_fixture(devnet_port: int):
-    proc = run_devnet(["poetry", "run", "starknet-devnet"], devnet_port)
-    yield f"http://localhost:{devnet_port}"
-    proc.kill()
+    with base_protostar_toml():
```
Comment (mkaput) [inline tests/integration/conftest.py:92]: why not using here the fixture you introduced in this PR?

File: tests/conftest.py (hunk 11 above)
Comment (mkaput) [inline tests/conftest.py:155]: also I wonder, shouldn't this be built into the `create_protostar_project` API?

File: tests/conftest.py (hunk 4 above)
Comment (mkaput) [reply tests/conftest.py:172]: agree

**Review by mkaput — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 4 above)
Comment (Arcticae) [reply tests/conftest.py:172]: this is needed for all existing tests to pass. extracting and using it all over the place seems less clean. most of existing e2e/integration tests would need to explicitly declare this fixture

Comment (mkaput) [reply tests/conftest.py:172]: but don't all tests use some other kinds of project-setup fixtures that could do this?

what is exactly failing *everywhere*?

Comment (Arcticae) [reply tests/conftest.py:172]: actually they don't, i.e. `run_cairo_test_runner_fixture` does not create a project

Comment (Arcticae) [reply tests/conftest.py:172]: I will circle back on this and look for alternatives, this was the easiest fix that i though of

File: protostar/testing/cheatcodes/declare_cheatcode.py (hunk 5 above)
Comment (Arcticae) [reply protostar/testing/cheatcodes/declare_cheatcode.py:65]: I will extract recognition of `.cairo` ending strings as cairo source files to `ProjectCompiler`. The second is out of scope of this PR and can be discussed/considered separately if we really want it, since it'd be modification of migrator's behavior.

File: tests/integration/conftest.py (hunk 13 above)
Comment (Arcticae) [reply tests/integration/conftest.py]: removed usage

File: tests/conftest.py (hunk 11 above)
Comment (Arcticae) [reply tests/conftest.py]: this contextmanager was moved to `run_cairo_test_runner` fixture which seems to resolve the issue

File: tests/conftest.py (hunk 12 above)
Comment (Arcticae) [reply tests/conftest.py]: renamed - i called it `empty_protostar_toml` since it concerns only the `.toml` file

File: tests/conftest.py (hunk 11 above)
Comment (Arcticae) [reply tests/conftest.py]: renamed

File: tests/benchmarks/test_benchmark.py (hunk 10 above)
Comment (Arcticae) [reply tests/benchmarks/test_benchmark.py]: i deleted benchmarking afterall, seems a bit silly to keep maintaining it, while it's unused.

File: tests/benchmarks/test_benchmark.py (hunk 9 above)
Comment (Arcticae) [reply tests/benchmarks/test_benchmark.py]: code was deleted

File: protostar/testing/environments/common_test_cheatcode_factory.py (hunk 7 above)
Comment (Arcticae) [reply protostar/testing/environments/common_test_cheatcode_factory.py]: done

File: protostar/testing/cheatcodes/declare_cheatcode.py (hunk 5 above)
Comment (Arcticae) [reply protostar/testing/cheatcodes/declare_cheatcode.py]: ^done

File: tests/benchmarks/test_benchmark.py (hunk 8 above)
Comment (Arcticae) [reply tests/benchmarks/test_benchmark.py]: done

File: tests/integration/conftest.py (hunk 6 above)
Comment (Arcticae) [reply tests/integration/conftest.py]: i moved this to `run_cairo_test_runner_fixture` - it needs it now

File: tests/conftest.py (hunk 4 above)
Comment (Arcticae) [reply tests/conftest.py]: fixed

File: protostar/testing/environments/test_execution_environment.py (hunk 1 above)
Comment (Arcticae) [reply protostar/testing/environments/test_execution_environment.py]: done

File: protostar/testing/environments/fuzz_test_execution_environment.py [hunk 14]
```diff
@@ -43,9 +44,10 @@ class FuzzTestExecutionResult(TestExecutionResult):
 
 
 class FuzzTestExecutionEnvironment(TestExecutionEnvironment):
-    def __init__(self, state: TestExecutionState):
-        super().__init__(state)
+    def __init__(self, state: TestExecutionState, project_compiler: ProjectCompiler):
+        super().__init__(state, project_compiler)
         self.initial_state = state
+        self._project_compiler = project_compiler
```
Comment (kasperski95) [inline protostar/testing/environments/fuzz_test_execution_environment.py:50]:
```suggestion
```
[Complexity] It looks like this property is not used.

File: tests/integration/conftest.py [hunk 15]
```diff
@@ -89,6 +89,7 @@ def assert_cairo_test_cases(
 
 @pytest.fixture(name="devnet_gateway_url", scope="session")
 def devnet_gateway_url_fixture(devnet_port: int):
+    # with base_protostar_toml():
```
Comment (kasperski95) [inline tests/integration/conftest.py:92]:
```suggestion
```

File: tests/integration/cheatcodes/declare/declare_contract_test.py [hunk 16]
```diff
@@ -23,3 +24,21 @@ async def test_declare_contract(run_cairo_test_runner: RunCairoTestRunnerFixture
...
+def test_declaring_contract_by_name(create_protostar_project):
+    with create_protostar_project() as protostar:
+        protostar.create_files(
+            {
+                "./src/main.cairo": Path(__file__).parent / "basic_contract.cairo",
+                "./tests/test_main.cairo": Path(__file__).parent
+                / "declaring_contract_by_name_test.cairo",
+            }
+        )
+        protostar.build_sync()
```
Comment (kasperski95) [inline tests/integration/cheatcodes/declare/declare_contract_test.py:39]:
```suggestion
```

File: tests/integration/cheatcodes/deploy_contract/deploy_contract_test.py [hunk 17]
```diff
@@ -35,3 +36,21 @@ async def test_deploy_contract(run_cairo_test_runner: RunCairoTestRunnerFixture)
...
+def test_deploying_contract_by_name(create_protostar_project):
+    with create_protostar_project() as protostar:
+        protostar.create_files(
+            {
+                "./src/main.cairo": Path(__file__).parent / "basic_contract.cairo",
+                "./tests/test_main.cairo": Path(__file__).parent
+                / "deploying_contract_by_name_test.cairo",
+            }
+        )
+        protostar.build_sync()
```
Comment (kasperski95) [inline tests/integration/cheatcodes/deploy_contract/deploy_contract_test.py:51]:
```suggestion
```

File: tests/integration/protostar_fixture.py (hunk 2 above)
Comment (Arcticae) [reply tests/integration/protostar_fixture.py:213]: black formats that file for me actually, dk really why

**Review by kasperski95 — APPROVED** [review]
LGTM.
I'd change the description of this PR from `fix` to `related` to avoid closing the issue before adding docs.

File: tests/integration/protostar_fixture.py (hunk 2 above)
Comment (mkaput) [reply tests/integration/protostar_fixture.py:213]: maybe outdated black version you have

**Review by mkaput — CHANGES_REQUESTED** [review]
All looks ok, but very much please, extract deletion of `test/benchmarks` and related changes to separate PR which will be merged earlier than this

**Comment by Arcticae** [comment]
@mkaput done
